### PR TITLE
Negative Intents

### PR DESCRIPTION
**Pull request by carlosfebres**
### Summary

  This PR introduces **Negative Intents**, a novel liquidity rebalancing mechanism for the crowd liquidity pool. Unlike traditional arbitrage-based rebalancing that relies on profitable opportunities,
  negative intents allow the system to incentivize rebalancing by offering fulfillers the ability to adjust their portfolio allocations while accepting a small, controlled loss (default 5%).

  ### Problem

  The crowd liquidity pool can experience token imbalances across chains. Traditional rebalancing relies on arbitrage opportunities, but these may not always exist or be profitable enough to incentivize
  immediate action, leading to:
  - Prolonged portfolio imbalances
  - Reduced capital efficiency
  - Inability to serve user intents optimally

  ### Solution

  Negative intents create a win-win rebalancing mechanism where:
  - **Crowd Liquidity Pool**: Gets rebalanced to maintain optimal token distribution
  - **Fulfillers**: Can rebalance their own portfolios while accepting a configured loss percentage

  ### How It Works

  1. **Intent Publication**: When rebalancing is needed, the system publishes an intent where:
     Fulfiller sends: 100 USDC on Chain A
     Fulfiller receives: 95 USDC on Chain B (5% loss)

  2. **Automated Fulfillment**: A Lit Action automatically fulfills the intent on behalf of the crowd liquidity pool

  3. **Event Monitoring**: The system tracks:
  - Transaction confirmation on destination chain
  - IntentProven event on source chain
  - Successful withdrawal execution

  4. **Completion**: After proving, rewards are withdrawn to complete the rebalancing

  ### Key Changes

  #### New Services
  - `NegativeIntentRebalanceProviderService`: Orchestrates the entire negative intent lifecycle
  - `NegativeIntentMonitorService`: Monitors intent execution and events
  - Refactored `LitActionService`: Centralized Lit Action execution with support for multiple action types

  #### Configuration
  ```typescript
  // liquidity-manager config
  {
    rebalancingPercentage: 0.05, // 5% loss accepted
    walletStrategies: {
     "crowd-liquidity": ["NegativeIntent", ...otherStrategies]
    }
  }
```

###  Type System

  - Added 'NegativeIntent' strategy type
  - Extended strategy context with NegativeIntentStrategyContext
  - Updated validation logic to accept negative reward intents

###  Testing

  ✅ Comprehensive unit tests for both services (16 tests total)
  ✅ Event monitoring with timeout handling
  ✅ Error scenarios and edge cases covered

###  Usage

  Enable negative intents by:
  1. Adding NegativeIntent to wallet strategies
  2. Configuring rebalancingPercentage (optional, defaults to 0.05)
  3. Deploying the negative intent Lit Action

### CODE REVIEW

File: src/liquidity-manager/services/liquidity-providers/negative-intents/negative-intent-rebalance.service.ts [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+  }
+
+  async getQuote(
+    tokenIn: TokenData,
+    tokenOut: TokenData,
+    swapAmount: number,
+    id?: string,
+  ): Promise<RebalanceQuote<'NegativeIntent'>> {
+    // Get the rebalancing percentage from config (default to 5% if not set)
+    const rebalancingPercentage = this.config.rebalancingPercentage || 0.05
```
Comment (StoyanD) [inline src/liquidity-manager/services/liquidity-providers/negative-intents/negative-intent-rebalance.service.ts:43]: magic number 0.05

File: src/eco-configs/eco-config.types.ts [hunk 2]
```diff
@@ -389,6 +389,8 @@ export interface LiquidityManagerConfig {
   walletStrategies: {
     [walletName: string]: Strategy[]
   }
+  // Percentage loss that fulfillers accept when rebalancing (e.g., 0.05 for 5%)
+  rebalancingPercentage?: number
```
Comment (StoyanD) [inline src/eco-configs/eco-config.types.ts:393]: 0.05 isnt a percent, consider using a percent value and dividing in the getter or switch to basis points

File: src/liquidity-manager/services/liquidity-providers/negative-intents/negative-intent-rebalance.service.ts [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+
+    // Parse amounts with proper decimals
+    const amountIn = parseUnits(swapAmount.toString(), tokenIn.balance.decimals)
```
Comment (StoyanD) [inline src/liquidity-manager/services/liquidity-providers/negative-intents/negative-intent-rebalance.service.ts:46]: is it safe to assume that swapAmount is always base 0 decimals?

File: src/liquidity-manager/services/liquidity-providers/negative-intents/negative-intent-rebalance.service.ts [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+    // Calculate the amount the fulfiller will receive (less than they spend)
+    const rebalancingBasisPoints = BigInt(Math.floor(rebalancingPercentage * 100_000))
```
Comment (StoyanD) [inline src/liquidity-manager/services/liquidity-providers/negative-intents/negative-intent-rebalance.service.ts:48]: seems simpler to just set `rebalancingBasisPoints` as the config value

File: src/liquidity-manager/services/liquidity-providers/negative-intents/negative-intent-rebalance.service.ts [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+    const amountOut = (amountIn * (100_000n - rebalancingBasisPoints)) / 100_000n
```
Comment (StoyanD) [inline src/liquidity-manager/services/liquidity-providers/negative-intents/negative-intent-rebalance.service.ts:49]: id move the calculations out into a util

File: src/liquidity-manager/services/liquidity-providers/negative-intents/negative-intent-rebalance.service.ts [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+          value: 0n,
+        },
+      ],
+    }
+
+    // Create the reward - fulfiller receives tokenOut (less than they spent)
+    const reward = {
+      creator: await this.kernelAccountClientService.getAddress(),
+      prover: intentSource.provers[0], // Use first available prover
+      deadline: BigInt(Math.floor(Date.now() / 1000) + 5_400), // 1.5 hours from now
```
Comment (StoyanD) [inline src/liquidity-manager/services/liquidity-providers/negative-intents/negative-intent-rebalance.service.ts:153]: utils method?

Comment (StoyanD) [reply src/liquidity-manager/services/liquidity-providers/negative-intents/negative-intent-rebalance.service.ts:153]: where do u get 1.5 hours from? is that a default somewhere

File: src/liquidity-manager/services/liquidity-providers/negative-intents/negative-intent-rebalance.service.ts [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+        properties: {
+          txHash,
+          intentHash,
+          id: quote.id,
+        },
+      }),
+    )
+
+    // Wait for transaction confirmation
+    await kernelClient.waitForTransactionReceipt({ hash: txHash })
```
Comment (StoyanD) [inline src/liquidity-manager/services/liquidity-providers/negative-intents/negative-intent-rebalance.service.ts:194]: should we store this rebalance in a db?

File: src/liquidity-manager/services/liquidity-providers/negative-intents/negative-intent-rebalance.service.ts [hunk 8]
```diff
@@ -0,0 +1,358 @@
...
+  private generateSalt(): Hex {
+    return `0x${randomBytes(32).toString('hex')}` as Hex
+  }
+
+  private getIntentSource(chainId: number) {
+    const intentSource = this.ecoConfigService
+      .getIntentSources()
+      .find((source) => source.chainID === chainId)
+
+    if (!intentSource) {
```
Comment (StoyanD) [inline src/liquidity-manager/services/liquidity-providers/negative-intents/negative-intent-rebalance.service.ts:282]: can u add error to eco-errors, feel like we throw this a lot

File: src/liquidity-manager/services/liquidity-providers/negative-intents/negative-intent-rebalance.service.ts [hunk 9]
```diff
@@ -0,0 +1,358 @@
...
+      throw error
+    }
+  }
+
+  private generateSalt(): Hex {
+    return `0x${randomBytes(32).toString('hex')}` as Hex
+  }
+
+  private getIntentSource(chainId: number) {
+    const intentSource = this.ecoConfigService
```
Comment (StoyanD) [inline src/liquidity-manager/services/liquidity-providers/negative-intents/negative-intent-rebalance.service.ts:278]: theres a method `getIntentSource(chainID: number)`

File: src/liquidity-manager/services/liquidity-providers/negative-intents/negative-intent-rebalance.service.ts [hunk 10]
```diff
@@ -0,0 +1,358 @@
...
+      const crowdLiquidityConfig = this.ecoConfigService.getCrowdLiquidity()
+      const publicClient = await this.publicClient.getClient(quote.tokenOut.chainId)
+
+      // Get fee data for the transaction
+      const [feeData, nonce] = await Promise.all([
+        this.getFeeData(publicClient),
+        publicClient.getTransactionCount({ address: crowdLiquidityConfig.pkp.ethAddress as Hex }),
+      ])
+
+      const transactionBase = { ...feeData, nonce, gasLimit: 1_000_000 }
```
Comment (StoyanD) [inline src/liquidity-manager/services/liquidity-providers/negative-intents/negative-intent-rebalance.service.ts:307]: magic 1_000_000

File: src/liquidity-manager/services/liquidity-providers/negative-intents/negative-intent-rebalance.service.ts [hunk 11]
```diff
@@ -0,0 +1,358 @@
...
+
+  private async getFeeData(publicClient: any) {
+    const [block, maxPriorityFeePerGas] = await Promise.all([
+      publicClient.getBlock(),
+      publicClient.estimateMaxPriorityFeePerGas(),
+    ])
+    const maxFeePerGas = block.baseFeePerGas! * 2n + maxPriorityFeePerGas
+
+    return {
+      type: 2,
```
Comment (StoyanD) [inline src/liquidity-manager/services/liquidity-providers/negative-intents/negative-intent-rebalance.service.ts:353]: whats a type 2? is there a reference definition

**Review by StoyanD — CHANGES_REQUESTED** [review]
looks good, added some comments on magic numbers, there is a reduce that will need to take decimals into account, and some other random stuff

File: src/liquidity-manager/services/liquidity-providers/negative-intents/negative-intent-rebalance.service.ts (hunk 3 above)
Comment (carlosfebres) [reply src/liquidity-manager/services/liquidity-providers/negative-intents/negative-intent-rebalance.service.ts:46]: No, the `swapAmount` is a number based on the token-in decimals.

File: src/liquidity-manager/services/liquidity-providers/negative-intents/negative-intent-rebalance.service.ts (hunk 4 above)
Comment (carlosfebres) [reply src/liquidity-manager/services/liquidity-providers/negative-intents/negative-intent-rebalance.service.ts:48]: It's simpler to implement but harder to change the value. Using `0.05` to represent 5% is much easier.

File: src/liquidity-manager/services/liquidity-providers/negative-intents/negative-intent-rebalance.service.ts (hunk 5 above)
Comment (carlosfebres) [reply src/liquidity-manager/services/liquidity-providers/negative-intents/negative-intent-rebalance.service.ts]: Done ✅  https://github.com/eco/eco-solver/pull/193/commits/2ad8d1e4a44fd170041e06fa3e22e7dc3f05bb02

File: src/liquidity-manager/services/liquidity-providers/negative-intents/negative-intent-rebalance.service.ts (hunk 6 above)
Comment (carlosfebres) [reply src/liquidity-manager/services/liquidity-providers/negative-intents/negative-intent-rebalance.service.ts:153]: I made a utils function and moved the deadline duration to the configuration.

File: src/liquidity-manager/services/liquidity-providers/negative-intents/negative-intent-rebalance.service.ts (hunk 8 above)
Comment (carlosfebres) [reply src/liquidity-manager/services/liquidity-providers/negative-intents/negative-intent-rebalance.service.ts:284]: Done ✅  https://github.com/eco/eco-solver/pull/193/commits/a5c2cd66d52f01395efccb351a36bb540e90b969

**Review by StoyanD — APPROVED** [review]
lets not merge this pr yet, 197 is in QA. we need to merge conflict resolve this pr once that one is in since it updates some of the same classes with normalization support
